modules/shortcodes/email-subscribe.php

```diff
@@ -0,0 +1,233 @@
+<?php
+/**
+ * WARNING: This file is distributed verbatim in Jetpack.
+ * There should be nothing WordPress.com specific in this file.
+ *
+ * @hide-in-jetpack
+ **/
+
+/**
+ * Class Jetpack_Email_Subscribe
+ * This class encapsulates shortcode for subscribing to a MailChimp list.
+ * It displays a simple signup form that gets an email address from the user and signs him for a list
+ * selected in "Sharing" section in calypso.
+ * Other Email services can be implemented as well in the future.
+ */
+class Jetpack_Email_Subscribe {
+
+	private static $shortcode = 'jetpack-email-subscribe';
+
+	private static $css_classname_prefix = 'jetpack-email-subscribe';
+
+	private static $option_name = 'jetpack_mailchimp';
+
+	private static $block_name = 'mailchimp';
+
+	private static $instance;
+
+	private static $version = '1.0';
+
+	private function __construct() {
+	}
+
+	/**
+	 * This follows a classic singleton pattern.
+	 *
+	 * @return Jetpack_Email_Subscribe|null
+	 */
+	public static function get_instance() {
+		// Do not load this at all if it's neither a WPCOM or a connected JP site.
+		if ( ! ( ( defined( 'IS_WPCOM' ) && IS_WPCOM ) || Jetpack::is_active() ) ) {
+			return null;
+		}
+
+		if ( ! self::$instance ) {
+			self::$instance = new self();
+			self::$instance->register_init_hook();
+		}
+
+		return self::$instance;
+	}
+
+	private function register_scripts_and_styles() {
+		wp_register_script( 'jetpack-email-subscribe', Jetpack::get_file_url_for_environment( '_inc/build/shortcodes/js/jetpack-email-subscribe.min.js', 'modules/shortcodes/js/jetpack-email-subscribe.js' ), array( 'jquery' ), self::$version );
+		wp_register_style( 'jetpack-email-subscribe', plugins_url( '/css/jetpack-email-subscribe.css', __FILE__ ), array(), self::$version );
+	}
+
+	private function register_init_hook() {
+		add_action( 'init', array( $this, 'init_hook_action' ) );
+		add_action( 'jetpack_options_whitelist', array( $this, 'filter_whitelisted_options' ), 10, 1 );
+		add_action( 'jetpack_blocks_to_register', array( $this, 'prevent_jetpack_register_block' ), 10, 1 );
+	}
+
+	public function prevent_jetpack_register_block( $blocks ) {
+		return array_filter( $blocks, array( $this, 'filter_out_this_block' ) );
+	}
+
+	public function filter_out_this_block( $block_name ) {
+		return ( $block_name !== self::$block_name );
+	}
+
+	public function filter_whitelisted_options( $options ) {
+		$options[] = self::$option_name;
+		return $options;
+	}
+
+	private function register_shortcode() {
+		add_shortcode( self::$shortcode, array( $this, 'parse_shortcode' ) );
+	}
+
+	private function register_gutenberg_block() {
+		if ( Jetpack_Gutenberg::is_gutenberg_available() ) {
+			register_block_type( 'jetpack/' . self::$block_name, array(
```

Comment:
There is a bug in Jetpack_Gutenberg  which makes Server-side rendered blocks impossible in Jetpack.

- `jetpack_register_block` makes `register_block_type` call too late for the route to SSR blocks to register. That is too late to WP_REST_Block_Renderer_Controller be aware of them.
- Calling `register_block_type` does make it register, but does not show the SDK blocks.
- calling both produces a notice in the logs.

Tne ONLY way of getting out of this conundrum is adding this filter OR registering a route for the SSR path seperately. I tried it and it broke MC, because MC requires registering routes in a special hook which makes no sense for Jetpack.

This is a catch-22. This filter is meant as a way of injecting jetpack_register_block

Reply:
> There is a bug in Jetpack_Gutenberg which makes Server-side rendered blocks impossible in Jetpack.

I think we have some instances of working server-side rendered blocks, e.g. [Contact Form](https://github.com/Automattic/jetpack/blob/master/modules/contact-form/grunion-contact-form.php).

This issue sounds familiar. I believe that we've run into it before, and I think we have a fix (or at least an established workaround) available. I'll need to study the code some more before being able to provide a better informed opinion.

Reply:
If there is a more elegant fix, I am happy to take it!

Reply:
So I just tried this

```diff
diff --git a/modules/shortcodes/email-subscribe.php b/modules/shortcodes/email-subscribe.php
index bc2564832..867059901 100644
--- a/modules/shortcodes/email-subscribe.php
+++ b/modules/shortcodes/email-subscribe.php
@@ -57,7 +57,7 @@ class Jetpack_Email_Subscribe {
 	private function register_init_hook() {
 		add_action( 'init', array( $this, 'init_hook_action' ) );
 		add_action( 'jetpack_options_whitelist', array( $this, 'filter_whitelisted_options' ), 10, 1 );
-		add_action( 'jetpack_blocks_to_register', array( $this, 'prevent_jetpack_register_block' ), 10, 1 );
+		//add_action( 'jetpack_blocks_to_register', array( $this, 'prevent_jetpack_register_block' ), 10, 1 );
 	}
 
 	public function prevent_jetpack_register_block( $blocks ) {
@@ -79,7 +79,7 @@ class Jetpack_Email_Subscribe {
 
 	private function register_gutenberg_block() {
 		if ( Jetpack_Gutenberg::is_gutenberg_available() ) {
-			register_block_type( 'jetpack/' . self::$block_name, array(
+			jetpack_register_block( self::$block_name, array(
 				'attributes' => array(
 					'title' => array(
 						'type' => 'string',
@@ -108,8 +108,7 @@ class Jetpack_Email_Subscribe {
 				),
 				'style' => 'jetpack-email-subscribe',
 				'render_callback' => array( $this, 'parse_shortcode' ),
-			) );
-			jetpack_register_block( self::$block_name, array(), array( 'available' => true ) );
+			), array( 'available' => true ) );
 		}
 	}
 
```

and the block seems to work both in the editor (and is availalble in the block picker), and in the frontend. Am I missing anything else that this would break?

Reply:
Yes. What breaks is Gutenberg on WPCOM. If you make that change in WPCOM codebase, this is what you get in Gutenlypso:

<img width="927" alt="zrzut ekranu 2019-01-03 o 17 28 08" src="https://user-images.githubusercontent.com/3775068/50649038-10e0b000-0f7d-11e9-9220-f4ea71c786af.png">

And since files have to be kept in sync, any fix has to be here too.

Reply:
Okay, I'll have a closer look at the WP.com side of things then :+1:

Reply:
Thank you! :)

Reply:
If we don't have a better solution for this, can this be documented in code comments? The combination of `prevent_jetpack_register_block()`, `register_gutenberg_block()`, and `jetpack_register_block()` is pretty confusing.

Reply:
Haven't been able to find a more generic fix yet (planning to get some other things out of the way first to make this easier to reason about, see pafL3P-cY-p2).

But yeah, definitely agree that this should be well-documented in code so we can cycle back and remove once we have a better solution.

Reply:
I'm confident that #11091 will fix this (and allow us to remove the workarounds that this PR introduced).